dnf/cli/commands/history.py

```diff
@@ -0,0 +1,506 @@
+                raise TransactionFileError(fn, str(e) + ".")
+
+        self._verify_toplevel_json()
+
+        rpms = self.replay_data.get("rpms", {})
+        if not isinstance(rpms, dict):
+            raise TransactionFileError(fn, _('Unexpected type of "rpms", object expected.'))
+
+        groups = self.replay_data.get("groups", {})
+        if not isinstance(groups, dict):
+            raise TransactionFileError(fn, _('Unexpected type of "groups", object expected.'))
+
+        environments = self.replay_data.get("environments", {})
+        if not isinstance(environments, dict):
+            raise TransactionFileError(fn, _('Unexpected type of "environments", object expected.'))
+
+        self._na_reason_cache = {}
+        errors = []
+
+        for na, pkg_data in rpms.items():
+            subj = hawkey.Subject(na)
+            nas = subj.get_nevra_possibilities(forms=[hawkey.FORM_NA])
+
+            if len(nas) != 1:
+                errors.append(TransactionError(_('Cannot parse name.arch for package "{na}".').format(na=na)))
+                continue
+
+            if isinstance(pkg_data, list):
+                # for installonly packages the data is a list of actions
+                for action_data in pkg_data:
+                    try:
+                        self._replay_pkg_action(nas[0], action_data)
+                    except TransactionError as e:
+                        errors.append(e)
+            else:
+                try:
+                    self._replay_pkg_action(nas[0], pkg_data)
+                except TransactionError as e:
+                    errors.append(e)
+
+        for group_id, group_data in groups.items():
+            try:
+                action = group_data["action"]
+
+                if action == "Install":
+                    try:
+                        pkg_types = libdnf.transaction.stringToCompsPackageType(group_data["package_types"])
+                    except libdnf.error.Error as e:
+                        errors.append(TransactionError(str(e)))
+                        continue
+
+                    self.base.group_install(group_id, pkg_types)
+                elif action == "Upgrade":
+                    self.base.group_upgrade(group_id)
+                elif action == "Removed":
+                    self.base.group_remove(group_id)
+                else:
+                    errors.append(TransactionError(
+                        _('Unexpected value of group action "{action}" for group "{group}".')
+                            .format(action=action, group=group_id)
+                    ))
+            except KeyError as e:
+                errors.append(TransactionError(
+                    _('Missing object key "{key}" under group "{group}".').format(key=e.args[0], group=group_id)
+                ))
+            except ValueError as e:
+                errors.append(TransactionError(str(e)))
+            except dnf.exceptions.CompsError as e:
```

Comment:
What about to be more flexible with --skip-broken option. This option is very important for any history transaction command because otherwise it will be impossible to use them on Fedora.